### PR TITLE
Reworked Flow Development Guide

### PR DESCRIPTION
**Pull request by knolleary**
This PR represents the working copy of the Flow Development Guide. It is still being written - the PR is to share progress and get feedback.

This PR contains the changes from the following PRs:
 - #106 Add Non-functional requirements chapter of flow-dev-guide 
 - #105 Add Project chapter of flow-dev-guide 
 - #102 Add Readability chapter of flow-dev-guide 
 - #101 Add Implementation chapter of flow-dev-guide 
 - #99 Add Designing chapter of flow-dev-guide 
 - #98 Add introduction of flow-dev-guide 

The content from those PRs has provided the starting point for the guide. Not all of the content from those PRs will be in the final guide.

The working draft of the guide contains the sections:

 - [Flow Dev guide index page](https://github.com/node-red/node-red.github.io/pull/163/files#diff-5ebc8fb7a6911bf3904d1866f4fee03f)
 - [Flow structure](https://github.com/node-red/node-red.github.io/pull/163/files#diff-afa24e59b4a1794de35bb2dba78c6356)
 - [Message design (not yet written)](https://github.com/node-red/node-red.github.io/pull/163/files#diff-4c7ae0d917ac49a704ee4e2ea89a7562)
 - ~Error handling (not yet written)~ see my comment below
 - [Documenting Flows](https://github.com/node-red/node-red.github.io/pull/163/files#diff-9c05749fcb854983786f355235a87e0a)

The current version of this branch can be viewed here: https://new.nodered.org/docs/developing-flows/

### CODE REVIEW

**Comment by knolleary** [comment]
I have removed the separate Handling Errors section of the guide as I realised we already have a section on how to do error handling in the main user guide.

Instead, I have added a smaller section to the Flow Structure guide that talks about how to layout the error handling and not about how to do error handling.

File: docs/developing-flows/documenting-flows.md [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+There is a balance to be considered here. The longer the label, the more space it needs in the flow. The shorter the label, the less information it can share.
+
+For some nodes, it might be appropriate to hide the label altogether to minimise the horizontal space it uses in the flow - giving more room to other nodes.
+
+Along with the label, nodes can also have a custom icon. For example, if you have a number of MQTT In nodes for different types of device, customising the icon to match the type of device could be helpful. This should be used with care as the icon is one of the main ways of identifying the type of a particular node
+
+Choosing good names for things applies just as much to the tabs and subflows used.
+
+It also very important for Link nodes. Without a name set, you have to use the Link node's internal ID when creating links between different tabs. That makes it hard to identify the right target node and mistakes can happen. If you consider the Link nodes as providing APIs between the different tabs, then a good choice of naming scheme will be needed.
```
Comment (dceejay) [inline docs/developing-flows/documenting-flows.md:68]: in particular naming things  with words like in and out can often be confusing...  is this in to a flow or in from a flow ?

File: docs/developing-flows/documenting-flows.md [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+  <p class="caption">Placeholder for image: Grouping nodes</p>
+</div>
+
+### Adding longer documentation
+
+All of the techniques discussed so far relate to the visual appearance of the flows. In order to add more in depth documentation, something more is needed.
+
+Every node, group and tab can have longer-form documentation added under Description tab in their edit dialog. This help can be formatted using Markdown and including lists, tables and links.
+
+This longer format of documentation is useful where more explanation is needed of a flow's purpose, or some more complex logic needs to be described.
```
Comment (dceejay) [inline docs/developing-flows/documenting-flows.md:113]: Where is this then visible.... ?

File: docs/developing-flows/flow-structure.md [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+Subflows let you create a new node in the palette whose internal implementation is described as a flow. You can then add new instances of the subflow wherever you would a normal node.
+
+<div style="width: 300px" class="figure">
+  <img src="images/placeholder.png" alt="Link nodes">
+  <p class="caption">Placeholder for image: Link nodes</p>
+</div>
+
+
+There are some important differences between the two approaches. Link nodes cannot be used in the middle of a flow, where messages are passed over the link and then return when the other flow completes. They can only be used to start or end a flow. They can also be connect to more than one other link node. This lets you pass messages out to multiple other flows, or have multiple flows pass messages into a single flow.
```
Comment (dceejay) [inline docs/developing-flows/flow-structure.md:49]: ... also be connect**ed** to ...

File: docs/developing-flows/flow-structure.md [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+
```
Comment (dceejay) [inline docs/developing-flows/flow-structure.md:50]: Links can also be used within a page to declutter if you have lots of wires feeding to or from a single point, or flying back from right to left.

File: docs/developing-flows/flow-structure.md [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+In the MQTT example, you could configure the node to publish to ${MY_TOPIC} and then add MY_TOPIC as a subflow property. When a user edits an individual instance they can then provide a custom value for MY_TOPIC.
+
+This pattern can be applied to any node configuration field that lets you enter the value directly. It doesn’t currently work for fields that are exposed as checkboxes or other custom UI elements.
+
+### Managing state information
+
+Another consideration is how to manage any state information in your flows. For example, keeping a count of how many messages pass through a flow, or the current state of an external sensor.
+
+Node-RED provides the Context system for managing state within the runtime. The context can be scoped to the same tab, subflow or made globally available.
```
Comment (dceejay) [inline docs/developing-flows/flow-structure.md:82]: globally available <-> available globally ?   2/10

File: docs/developing-flows/flow-structure.md [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+
```
Comment (dceejay) [inline docs/developing-flows/flow-structure.md:83]: Should we also mention persistence over restart here as well  (or lack of by default) ? link to other docs etc ?

File: docs/developing-flows/documenting-flows.md (hunk 1 above)
Comment (dceejay) [inline docs/developing-flows/documenting-flows.md:68]: for example labelling things with in and out can still be confusing . Is this in to a flow or in from another flow.

**Comment by kazuhitoyokoi** [comment]
#107 is part of the node development guideline. Therefore, it can be removed from this flow development guideline.

**Comment by knolleary** [comment]
@kazuhitoyokoi you are right - #107 is **not** included in this PR. I listed it by mistake.


To make it easier to review this guide, I have set https://new.nodered.org - which is the `flowdev` branch.

The flow developer guide can be found here: https://new.nodered.org/docs/developing-flows/

File: docs/developing-flows/flow-structure.md (hunk 6 above)
Comment (knolleary) [reply docs/developing-flows/flow-structure.md:83]: I plan to do a separate pass through the doc adding links to relevant sections.

**Comment by knolleary** [comment]
Made some minor updates to reflect the comments from @dceejay  - have updated https://new.nodered.org/docs/developing-flows/ with the latest

**Comment by knolleary** [comment]
I have now added the Message section to the guide - https://new.nodered.org/docs/developing-flows/

This completes my first full pass through of the guide.

**Question:** is there any more content from the Hitatchi guide that should be added to the guide?

I would like to remove the extra Hitachi content before opening it for wider review with the community.

**Comment by kazuhitoyokoi** [comment]
Thank you for your confirmation. I checked the website. This pull request contains all of the contents from Hitachi.

**Comment by knolleary** [comment]
@kazuhitoyokoi I need to clarify what I plan to do as I'm not sure from your reply if it is clear.

Yes this PR contains all of the content from Hitachi. But the plan is to delete all of the pages from Hitachi because the information has been rewritten and included in the three sections I have been working on.

image

I want to make sure you are happy that the three sections I have written cover everything you think it should and that we are okay to delete the hitachi pages.

File: docs/developing-flows/message-design.md [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+layout: docs-developing-flows
+toc: toc-developing-flows.html
+title: Message design
+slug: message design
+---
+
+The messages that pass through a flow are plain JavaScript objects that can have
+properties set on them.
+
+They usually have a payload property - this is the default property that most nodes
```
Comment (dceejay) [inline docs/developing-flows/message-design.md:11]: do we want to emphasise payload - eg like below - or maybe just bold

File: docs/developing-flows/message-design.md [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+will work with.
+
+For more information about messages in Node-RED you should read the [Working with messages](/docs/user-guide/messages)
+section of the user guide.
+
+This section looks at some of the choices you need to make when deciding how to
+stucture the messages in your flows.
+
+### Working with the payload
```
Comment (dceejay) [inline docs/developing-flows/message-design.md:20]: should this be `msg.payload`  to link to the `msg.topic` heading below

File: docs/developing-flows/message-design.md [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+
+<div style="width: 300px" class="figure">
+  <img src="images/placeholder.png" alt="Link nodes">
+  <p class="caption">Placeholder for image: example change node to save msg.id</p>
+</div>
+
+
+This highlights an important principle: nodes should not modify or remove properties on messages that are not related to their functionality.
+
+For example, in most cases, a Function node should send on the same message object it received rather than create a new message object.
```
Comment (dceejay) [inline docs/developing-flows/message-design.md:44]: just a Function node ? (Ok I get for most users yes... but...)

File: docs/developing-flows/message-design.md [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+This may be convenient to keep the data together, but it can also lead to a lot of moving properties around as later nodes expect to operate on `msg.payload` and not a property underneath it.
+
+A different approach, as seen by the Twitter node, is to put the most 'interesting' information into the payload, in this case the text of a tweet, and put the complete metadata the API also provides into a separate `msg.tweet` property.
+
+There is not a single right answer to how to structure the message, but it should focus on how the node or flow is going to be used in the most common cases.
+
+As with programming in general, the choice of good property names is also important. They should be self-describing to help with later debugging and understanding of the flow. For example, `msg.temperature` is much more understandable than `msg.t`.
+
+
+##
```
Comment (dceejay) [inline docs/developing-flows/message-design.md:86]: supefluous

File: docs/developing-flows/message-design.md [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+
+This may be convenient to keep the data together, but it can also lead to a lot of moving properties around as later nodes expect to operate on `msg.payload` and not a property underneath it.
+
+A different approach, as seen by the Twitter node, is to put the most 'interesting' information into the payload, in this case the text of a tweet, and put the complete metadata the API also provides into a separate `msg.tweet` property.
+
+There is not a single right answer to how to structure the message, but it should focus on how the node or flow is going to be used in the most common cases.
+
+As with programming in general, the choice of good property names is also important. They should be self-describing to help with later debugging and understanding of the flow. For example, `msg.temperature` is much more understandable than `msg.t`.
+
+
```
Comment (dceejay) [inline docs/developing-flows/message-design.md:85]: should we mention any properties NOT to use ? eg id or parts

File: docs/developing-flows/documenting-flows.md [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+### Naming nodes
+
+Most nodes have a `name` property that can be used to customise the label they display in the workspace. This should be used to properly label the key points of a flow.
+
+For example, if a Change node has a single rule that sets `msg.payload` to the current time, its default label will be `set msg.payload`. That helps somewhat, but it doesn't reveal the full purpose of the node. A name of `Get current time` would be much clearer.
+
+There is a balance to be considered here. The longer the label, the more space it needs in the flow. The shorter the label, the less information it can share.
+
+For some nodes, it might be appropriate to hide the label altogether to minimise the horizontal space it uses in the flow - giving more room to other nodes.
+
```
Comment (dceejay) [inline docs/developing-flows/documenting-flows.md:63]: maybe an image of the "appearance tab" in config to show how to get to shorten the name

File: docs/developing-flows/documenting-flows.md [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+### Adding port labels
+
+If a node has multiple outputs it can be hard to follow the logic if it is not clear on what condition a message may be sent from a particular output.
+
+This is where adding port labels can help document the intended logic.
+
+For example, the Switch node provides default labels for its outputs that are shown when the mouse hovers over them. They can help quickly identify the purpose of each branch in the flow.
+
+Whilst the default labels may be sufficient in the context of the flow itself, it is also possible to custom labels to provide more detailed information.
```
Comment (dceejay) [inline docs/developing-flows/documenting-flows.md:78]: ... also possible for custom ...

File: docs/developing-flows/documenting-flows.md [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+By indenting the flows on the page, you can indicate an implied grouping of the different components.
+
+<div style="width: 600px" class="figure">
+  <img src="images/placeholder.png" alt="Link nodes">
+  <p class="caption">Placeholder for image: Indenting with Comments</p>
+</div>
+
+### Grouping nodes
+
+A more explicit arrangement of the flows can be achieved by grouping related nodes together.
```
Comment (dceejay) [inline docs/developing-flows/documenting-flows.md:98]: Link to grouping docs ? Or cut/paste blog post info to here ?

**Comment by dceejay** [comment]
In intro para 3 - I think it is also worth adding a link to the YouTube introduction channel  - as that is a good resource.
Do we want to add links to other parts of the docs - eg re context ? and or cookbook ?

File: docs/developing-flows/message-design.md (hunk 9 above)
Comment (knolleary) [reply docs/developing-flows/message-design.md:44]: The Function node is the only node a user can explicitly create a new message object which is why it is singled out here.

**Comment by kazuhitoyokoi** [comment]
@knolleary I misunderstood your question. Because today is a holiday in Japan, I will answer it on Monday after checking a little with Hitachi colleagues. Please wait a moment. 🙏

**Comment by knolleary** [comment]
No rush - want to make sure we get this right.

I hope this plan does not come as a surprise - I have been saying since the start of reviewing the flow dev guide that I felt the best approach was to take the information from the Hitachi content and rework it into a slightly different shape.

**Comment by kazuhitoyokoi** [comment]
@knolleary Thank you for waiting. Today, I talked about it with Hitachi colleagues. Because three sections contain key points, we agree with your decision. You can delete the Hitachi pages.

**Comment by knolleary** [comment]
Thanks @kazuhitoyokoi 

And please do remember, this is not a 'finished' guide - we can add extra content at any time. So if there are other topics you think it should cover in the future, then open an issue/pr and we can discuss.

**Comment by kazuhitoyokoi** [comment]
@knolleary I appreciate your consideration. Okay, we will suggest a discussion if we have feedback from the Japanese community. 😁

**Comment by knolleary** [comment]
I've updated flow dev guide to remove the extra pages. I'm going to share a link for the wider community to provide some feedback before we publish.

In terms of content, we still need to:

1. create the images
2. ensure consistent formatting when referencing nodes or message properties
3. add links to related parts of the docs and videos

**Comment by knolleary** [comment]
Going to merge this content as it currently stands.

I'm sure there's more to add, particularly around Performance, but I'd rather get this material published and available.
